### PR TITLE
Show only skills that the hero doesn't already have in battle only skill selection list;

### LINKED ISSUE #7534: Battle only mode and maybe future editor; possibility to select same skill more than once
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.
- [X] I've checked that this issue is reproducible on the [**latest snapshot build**](https://ihhub.github.io/fheroes2/INSTALL.html#snapshots-latest-builds).

### Platform

Windows

### Describe the bug

In the battle only mode it is possible to select the same skill, at the same level or different levels, more than once in the skill selection dialog. The future editor uses these dialogs so this issue will need to be solved for that too.

image

### Save file

Prohibit a skill from being applied more than once.

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by petar-stamenkov**
Fixes #7534

### CODE REVIEW

File: src/fheroes2/dialog/dialog_selectitems.cpp [hunk 1]
```diff
@@ -372,7 +382,7 @@ class SelectEnumSecSkill : public SelectEnum
 
     void RedrawItem( const int & index, int32_t dstx, int32_t dsty, bool current ) override
     {
-        const Skill::Secondary skill( 1 + index / 3, 1 + ( index % 3 ) );
+        const Skill::Secondary skill( getSkillFromListIndex( index ), getLevelFromListIndex( index ) );
```
Comment (Districh-ru) [inline src/fheroes2/dialog/dialog_selectitems.cpp:385]: Hi, @petar-stamenkov, we can do the same change here:
https://github.com/ihhub/fheroes2/blob/1145386e98b2da5490995c6a4bcede9b304ec371/src/fheroes2/dialog/dialog_selectitems.cpp#L417

Comment (petar-stamenkov) [reply src/fheroes2/dialog/dialog_selectitems.cpp:385]: Hi @Districh-ru, thanks for the suggestion! I've updated it.

File: src/fheroes2/dialog/dialog_selectitems.cpp [hunk 2]
```diff
@@ -372,23 +382,26 @@ class SelectEnumSecSkill : public SelectEnum
...
 
-Skill::Secondary Dialog::selectSecondarySkill( const int skillId /* = Skill::Secondary::UNKNOWN */ )
+Skill::Secondary Dialog::selectSecondarySkill( const Heroes & hero, const int skillId /* = Skill::Secondary::UNKNOWN */ )
 {
     std::vector<int> skills( static_cast<size_t>( MAXSECONDARYSKILL * 3 ), 0 );
 
     std::iota( skills.begin(), skills.end(), 0 );
+    skills.erase( std::remove_if( skills.begin(), skills.end(),
+                                  [&hero]( const int listIndex ) { return hero.HasSecondarySkill( SelectEnumSecSkill::getSkillFromListIndex( listIndex ) ); } ),
+                  skills.end() );
```
Comment (Districh-ru) [inline src/fheroes2/dialog/dialog_selectitems.cpp:404]: @petar-stamenkov, what do you think about slightly simplifying this logic?
Currently we make a vector filled with `0`. Then we fill it with integers from 0 to (MAXSECONDARYSKILL * 3 -1). And then we erase some integers, checking if hero has this skill.
We can fill the vector with the needed integers, doing the suggested check when filling, in example:
```cpp
    std::vector<int> skills;
    skills.reserve( MAXSECONDARYSKILL * 3 );

    for ( int i = 0; i < MAXSECONDARYSKILL * 3; ++i ) {
        if ( !hero.HasSecondarySkill( SelectEnumSecSkill::getSkillFromListIndex( i ) ) ) {
            skills.push_back( i );
        }
    }
```

File: src/fheroes2/dialog/dialog_selectitems.cpp (hunk 1 above)
Comment (Districh-ru) [reply src/fheroes2/dialog/dialog_selectitems.cpp:385]: > Hi @Districh-ru, thanks for the suggestion! I've updated it.

@petar-stamenkov, thanks for updating the `ActionListPressRight()` with `getSkillFromListIndex()` and `getLevelFromListIndex()`! I missed it.
Could you please also update the `return` in `Dialog::selectSecondarySkill()`, line 417. :)

Comment (petar-stamenkov) [reply src/fheroes2/dialog/dialog_selectitems.cpp:385]: sorry, fixed now :)

File: src/fheroes2/dialog/dialog_selectitems.cpp (hunk 2 above)
Comment (petar-stamenkov) [reply src/fheroes2/dialog/dialog_selectitems.cpp]: That looks cleaner, thanks. I've updated it.

File: src/fheroes2/dialog/dialog_selectitems.cpp [hunk 3]
```diff
@@ -372,23 +382,28 @@ class SelectEnumSecSkill : public SelectEnum
...
+        fheroes2::SecondarySkillDialogElement( Skill::Secondary( getSkillFromListIndex( index ), getLevelFromListIndex( index ) ), Heroes() ).showPopup( Dialog::ZERO );
     }
 };
 
-Skill::Secondary Dialog::selectSecondarySkill( const int skillId /* = Skill::Secondary::UNKNOWN */ )
+Skill::Secondary Dialog::selectSecondarySkill( const Heroes & hero, const int skillId /* = Skill::Secondary::UNKNOWN */ )
 {
-    std::vector<int> skills( static_cast<size_t>( MAXSECONDARYSKILL * 3 ), 0 );
+    std::vector<int> skills;
+    skills.reserve( MAXSECONDARYSKILL * 3 );
```
[the remark this answers is not in the document]
Comment (Districh-ru) [reply src/fheroes2/dialog/dialog_selectitems.cpp:400]: We can do a static casting here to please ClangTidy: `static_cast<size_t>( MAXSECONDARYSKILL * 3 )`.
Sorry, I forgot it in my example.

**Review by Districh-ru — APPROVED** [review]
Well done, @petar-stamenkov!

**Review by ihhub — APPROVED** [review]

**Comment by ihhub** [comment]
@petar-stamenkov , huge thanks for handling this issue!

**Pull request merged**